### ISSUE TITLE
Tensorflow ExponentialDecay is mis-implemented

### ISSUE THREAD

**Issue by gehlhaarPfizer** [deepchem/deepchem#3991]
This bug is in the latest released version of DeepChem (2.8.0).

The ExponentialDecay learning rate scheduler is mis-implemented in the code for TensorFlow. The platform-specific learning rate scheduler built at initialization, is calling the generated TensorFlow scheduler with the current global step -- which is always zero, because this happens once at initialization. This locks the learning rate at the initial learning rate. 

Incorrect code:
```
    def _create_tf_tensor(self, global_step):
        import tensorflow as tf
        return tf.keras.optimizers.schedules.ExponentialDecay(
            initial_learning_rate=self.initial_rate,
            decay_rate=self.decay_rate,
            decay_steps=self.decay_steps,
            staircase=self.staircase)(global_step)
```

Correct code:

```
    def _create_tf_tensor(self, global_step):
        import tensorflow as tf
        return tf.keras.optimizers.schedules.ExponentialDecay(
            initial_learning_rate=self.initial_rate,
            decay_rate=self.decay_rate,
            decay_steps=self.decay_steps,
            staircase=self.staircase)
```

**Comment by gehlhaarPfizer** [comment]
Honestly, I don’t know. I have only ever used Keras/Tensorflow. I literally have never used PyTorch. The vast majority of the examples of the kind I am interested in use Keras models (e.g. GraphConvModel).  I hope the models are all getting updated/replaced to use the desired engine. In particular I don’t know if there is a model I should be using instead of GraphConvModel that is based on PyTorch.

From: Bharath Ramsundar ***@***.***>
Sent: Friday, May 31, 2024 11:17 AM
To: deepchem/deepchem ***@***.***>
Cc: Gehlhaar, Daniel Kurt ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [deepchem/deepchem] Tensorflow ExponentialDecay is mis-implemented (Issue #3991)


Thank you for the bug report! The TensorFlow code is in maintenance mode now and we are swapping our primary backend over to Torch (since Torch is maintained by the Linux foundation, https://www.linuxfoundation.org/blog/blog/welcoming-pytorch-to-the-linux-foundation<https://urldefense.com/v3/__https:/www.linuxfoundation.org/blog/blog/welcoming-pytorch-to-the-linux-foundation__;!!H9nueQsQ!7pV7YrhXeYW2GRIxpMv_L6ht_-gmTZIyPUNDesg31AoKbGMF7CdOG-ewqaTCYyiNTXjs8uinF01aaM5-qcQxRDJqMSY$>, it is a better base for a long term stable scientific code). If possible, I'd encourage you to swap to Torch models (we will make this default from 2.9.0 onwards).

Is the decay broken on Torch as well?

**Comment by gehlhaarPfizer** [comment]
OK thanks. We are just getting started with DeepChem (we have prior experience with directly using Keras/TF). I’m not a data scientist, I do computational chemistry stuff, so I am feeling my way along here. We can experiment with some of the PyTorch models to get our feet wet.

From: Bharath Ramsundar ***@***.***>
Sent: Friday, May 31, 2024 11:28 AM
To: deepchem/deepchem ***@***.***>
Cc: Gehlhaar, Daniel Kurt ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [deepchem/deepchem] Tensorflow ExponentialDecay is mis-implemented (Issue #3991)


We are working on a 1-1 conversion of GraphConvModel (CC @NimishaDey<https://urldefense.com/v3/__https:/github.com/NimishaDey__;!!H9nueQsQ!7xpY6uXeEwoCK5-bY5QvOAdb6hPa0H70hZXgJYrzk_hqOXOYbADKygac1lE4NhiCil1YoegMl1b-3JSUVW9AMtLksRY$>) that should be done soon. We will update this thread later once that converted model is available. We are working on converting the tutorials/docs to use Torch. That work is underway and should be completed hopefully by the end of the year.

**Comment by gehlhaarPfizer** [comment]
I have verified that the PyTorch implementation is correct and functional.
